src/iceberg/delete_file_index.cc

```diff
@@ -0,0 +1,769 @@
+      continue;
+    }
+
+    if (AllNull(data_file.null_value_counts, data_file.value_counts, field_id,
+                is_required) &&
+        AllNonNull(wrapped_delete_file->null_value_counts, field_id, is_required)) {
+      return false;  // Data is all null, delete has no nulls - cannot match
+    }
+
+    if (AllNull(wrapped_delete_file->null_value_counts, wrapped_delete_file->value_counts,
+                field_id, is_required) &&
+        AllNonNull(data_file.null_value_counts, field_id, is_required)) {
+      return false;  // Delete is all null, data has no nulls - cannot match
+    }
+
+    if (!check_ranges) {
+      continue;
+    }
+
+    // Check range overlap
+    auto data_lower_it = data_file.lower_bounds.find(field_id);
+    auto data_upper_it = data_file.upper_bounds.find(field_id);
+    if (data_lower_it == data_file.lower_bounds.cend() || data_lower_it->second.empty() ||
+        data_upper_it == data_file.upper_bounds.cend() || data_upper_it->second.empty()) {
+      continue;  // Missing bounds, assume may match
+    }
+
+    auto delete_lower = delete_file.LowerBound(field_id);
+    auto delete_upper = delete_file.UpperBound(field_id);
+    if (!delete_lower.has_value() || !delete_upper.has_value()) {
+      continue;  // Missing bounds, assume may match
+    }
+
+    // Convert data bounds
+    auto primitive_type = checked_pointer_cast<PrimitiveType>(field.type());
+    ICEBERG_ASSIGN_OR_RAISE(auto data_lower,
+                            Literal::Deserialize(data_lower_it->second, primitive_type));
+    ICEBERG_ASSIGN_OR_RAISE(auto data_upper,
+                            Literal::Deserialize(data_upper_it->second, primitive_type));
+
+    if (!RangesOverlap(data_lower, data_upper, delete_lower->value().get(),
+                       delete_upper->value().get())) {
+      return false;  // Ranges don't overlap - cannot match
+    }
+  }
+
+  return true;
+}
+
+// PositionDeletes implementation
+
+Status PositionDeletes::Add(ManifestEntry&& entry) {
+  ICEBERG_PRECHECK(entry.sequence_number.has_value(),
+                   "Missing sequence number for position delete: {}",
+                   entry.data_file->file_path);
+  files_.emplace_back(std::move(entry));
+  indexed_ = false;
+  return {};
+}
+
+std::vector<std::shared_ptr<DataFile>> PositionDeletes::Filter(int64_t seq) {
+  IndexIfNeeded();
+
+  auto iter = std::ranges::lower_bound(seqs_, seq);
+  if (iter == seqs_.end()) {
+    return {};
+  }
+  return files_ | std::views::drop(iter - seqs_.begin()) |
+         std::views::transform(&ManifestEntry::data_file) |
+         std::ranges::to<std::vector<std::shared_ptr<DataFile>>>();
+}
+
+std::vector<std::shared_ptr<DataFile>> PositionDeletes::ReferencedDeleteFiles() {
+  IndexIfNeeded();
+  return files_ | std::views::transform(&ManifestEntry::data_file) |
+         std::ranges::to<std::vector<std::shared_ptr<DataFile>>>();
+}
+
+void PositionDeletes::IndexIfNeeded() {
+  if (indexed_) {
```

Comment:
The Java implementation uses double-checked locking with synchronized blocks:

```
  // Java
  private void indexIfNeeded() {
    if (buffer != null) {
      synchronized (this) {
        if (buffer != null) {
          this.files = indexFiles(buffer);
          this.seqs = indexSeqs(files);
          this.buffer = null;
        }
      }
    }
  }
```

  The C++ version has no synchronization

Reply:
Same comment as above.